src/osmdata-data_frame.cpp

```diff
@@ -0,0 +1,267 @@
+ **                                                                    **
+ ************************************************************************
+ ************************************************************************/
+
+
+//' get_osm_relations
+//'
+//' Return a dual Rcpp::DataFrame containing all OSM relations.
+//'
+//' @param rels Pointer to the vector of Relation objects
+//' @param unique_vals Pointer to a UniqueVals object containing std::sets of all
+//'       unique IDs and keys for each kind of OSM object (nodes, ways, rels).
+//'
+//' @return A dual Rcpp::DataFrame with the tags and metadata of the relations.
+//' 
+//' @noRd 
+Rcpp::DataFrame osm_df::get_osm_relations (const Relations &rels, 
+        const UniqueVals &unique_vals)
+{
+
+    const unsigned int nmp = rels.size ();
+
+    size_t ncol = unique_vals.k_rel.size ();
+    std::vector <std::string> rel_ids; 
+    rel_ids.reserve (nmp);
+
+    Rcpp::CharacterMatrix kv_mat (Rcpp::Dimension (nmp, ncol));
+    unsigned int count = 0;
+
+    for (auto itr = rels.begin (); itr != rels.end (); ++itr)
+    {
+        if (count % 1000 == 0)
+            Rcpp::checkUserInterrupt ();
+
+        rel_ids.push_back (std::to_string (itr->id));
+
+        osm_convert::get_value_mat_rel (itr, unique_vals, kv_mat, count++);
+    }
+
+    Rcpp::DataFrame kv_df;
+    if (nmp > 0)
+    {
+        kv_mat.attr ("names") = unique_vals.k_rel;
+        kv_mat.attr ("dimnames") = Rcpp::List::create (rel_ids, unique_vals.k_rel);
+        kv_df = osm_convert::restructure_kv_mat (kv_mat, false);
+    } else
+        kv_df = R_NilValue;
+
+    rel_ids.clear ();
+
+    return kv_df;
+}
+
+//' get_osm_ways
+//'
+//' Store key-val pairs for OSM ways as a list/data.frame
+//'
+//' @param kv_df Pointer to Rcpp::DataFrame to hold key-value pairs
+//' @param way_ids Vector of <osmid_t> IDs of ways to trace
+//' @param ways Pointer to all ways in data set
+//' @param unique_vals pointer to all unique values (OSM IDs and keys) in data set
+//' 
+//' @noRd 
+Rcpp::DataFrame osm_df::get_osm_ways (
+        const std::set <osmid_t> &way_ids, const Ways &ways,
+        const UniqueVals &unique_vals)
+{
+
+    size_t nrow = way_ids.size (), ncol = unique_vals.k_way.size ();
+    std::vector <std::string> waynames;
+    waynames.reserve (nrow);
+
+    Rcpp::CharacterMatrix kv_mat (Rcpp::Dimension (nrow, ncol));
+    std::fill (kv_mat.begin (), kv_mat.end (), NA_STRING);
+    unsigned int count = 0;
+    for (auto wi = way_ids.begin (); wi != way_ids.end (); ++wi)
+    {
+        if (count % 1000 == 0)
+            Rcpp::checkUserInterrupt ();
+
+        waynames.push_back (std::to_string (*wi));
+
+        auto wj = ways.find (*wi);
+        osm_convert::get_value_mat_way (wj, unique_vals, kv_mat, count);
+        count++;
+    }
+
+    Rcpp::DataFrame kv_df = R_NilValue;
+    if (way_ids.size () > 0)
+    {
+        kv_mat.attr ("names") = unique_vals.k_way;
```

Comment:
```suggestion
```
`names` attribute is the same as the second dimension of the `dimnames` assigned in the following line